### PR TITLE
Fix WithIngressClass parameter resolution for Kubernetes publish

### PR DESCRIPTION
**Pull request by mitchdenny**
## Description

When an ingress or gateway resource referenced a `ParameterResource` via `WithIngressClass`, `WithHostname`, `WithTls`, etc., the generated Kubernetes YAML rendered the literal format string `"{0}"` for unresolved parameters instead of a Helm template expression. This made the published Helm chart unusable unless the parameter happened to have a value at publish time. Reproduced from a bug shown on AspireFridays by @robrich:

```csharp
var ingressClass = builder.AddParameter("ingressclass");
ingress.WithIngressClass(ingressClass);
```

produced:

```yaml
spec:
  ingressClassName: "{0}"     # should be a Helm template reference
```

### Why the bug existed

`KubernetesEnvironmentResource.ResolveExpressionAsync` was the single helper used to resolve every `ReferenceExpression` consumed by the Kubernetes publisher (ingress class, hostname, TLS, gateway listeners, etc.). It called `ReferenceExpression.GetValueAsync` and, on `MissingParameterValueException`, returned `expression.Format` verbatim as a "fallback":

```csharp
catch (MissingParameterValueException)
{
    return expression.Format;   // "{0}" for a single ParameterResource
}
```

`expression.Format` is the raw composite format string (e.g. `"{0}"` for one parameter, `"{0}-{1}"` for two). It only resembles a useful value for expressions that contain no placeholders, which is essentially never the case for parameter-driven values. The docstring claimed the helper returned the parameter name, but the implementation never did — it just emitted the format string and let it land in the published YAML.

This was harmless for the original callers (TLS bootstrap and gateway address discovery in pipeline steps) because those run at deploy time, where parameters already have values and the catch block is never hit. The bug only surfaced once the same helper was reused for publish-time manifest generation, where missing values are the normal case.

### How the fix works

`ResolveExpressionAsync` is now parameter-aware and runs at publish time. When the expression contains unresolved `ParameterResource`s, the fix:

1. Walks each `IValueProvider` in the expression individually instead of returning the raw format string.
2. For each unresolved `ParameterResource`, substitutes a Helm template reference of the form `{{ .Values.parameters.<owningResource>.<paramName> }}` (or `secrets.` for secret parameters), scoped to the owning ingress/gateway resource so multiple resources can reuse the same parameter name without colliding.
3. Captures the parameter in `CapturedHelmValues` so two downstream consumers get the right data:
   - `HelmDeploymentEngine` writes the resolved parameter value into the deploy-time override file Aspire passes to `helm install`/`helm upgrade`.
   - `values.yaml` gets a matching placeholder entry via the new `EnsureCapturedHelmValuePlaceholders` pass in `KubernetesPublishingContext`, so `helm template` against the published chart does not render `<no value>` for users who deploy outside of Aspire.
4. Re-applies `string.Format` with the substituted segments, preserving any literal text from the original expression (e.g. an expression like `"prefix-{0}-suffix"` becomes `"prefix-{{ .Values.parameters.ingress.foo }}-suffix"`).

Pipeline-time callers (TLS bootstrap, gateway address discovery) keep their original behavior through a renamed static helper `ResolveExpressionAtDeployTimeAsync`. Those paths run after parameters are guaranteed to be resolved, so a missing-value fallback there is acceptable (and, again, never hit in practice).

### Testing notes

Added two regression tests in `KubernetesIngressTests.cs` covering the parameter and missing-parameter cases. The missing-parameter test needs `Pipeline:ClearCache=true` plus a per-run unique parameter name because the deployment state file at `~/.aspire/deployments/<sha>/<env>.json` persists across local test runs and would otherwise auto-resolve the parameter from a previous run, masking the failure.

Fixes # (issue)

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No

### CODE REVIEW

**Comment by mitchdenny** [comment]
## Deployment test run analysis (`/deployment-test` run [#26348648275](https://github.com/microsoft/aspire/actions/runs/26348648275))

Four deployment tests failed on this PR. All four are **pre-existing failures unrelated to this change** — they fail identically on the most recent scheduled `Deployment Tests` runs on `main` (#26323258106, #26268417397, #26205530633, #26141293553, #26076159680).

Notably, every other Kubernetes/Helm/AKS deployment test passed, including the C# variant of the cert-manager AKS scenario — only the TypeScript variant failed.

Filed tracking issues (issue refs use `microsoft/aspire#NNNN` syntax so this PR's merge doesn't auto-close them):

| Test | Issue |
|------|-------|
| `FrontDoorDeploymentTests` (CLI prompt regression) | microsoft/aspire#17435 |
| `TypeScriptAzureContainerAppJobDeploymentTests` (TS apphost hang) | microsoft/aspire#17436 |
| `TypeScriptJavaScriptHostingDeploymentTests` (TS apphost early ERR) | microsoft/aspire#17437 |
| `AksAzureKubernetesEnvironmentCertManagerTypeScriptDeploymentTests` (TS template `apphost.ts` path mismatch) | microsoft/aspire#17438 |

**Comment by mitchdenny** [comment]
Re: feedback from @davidfowl that "the parameter processing looks sus — either you are delegating to helm or processing it in proc":

The Kubernetes publisher does both, intentionally and consistently. Every `ParameterResource` consumed by the publisher already follows this pattern (env vars, ports, image, conditional branches — see `KubernetesResource.AllocateParameter` in `src/Aspire.Hosting.Kubernetes/KubernetesResource.cs`):

1. The published manifest gets a Helm template reference: `{{ .Values.parameters.<resource>.<param> }}` (or `secrets.<...>` for secret parameters).
2. The parameter is captured into `KubernetesEnvironmentResource.CapturedHelmValues`.
3. At deploy time, `HelmDeploymentEngine` resolves the captured parameters in-proc and writes them into a values override file passed to `helm install`/`helm upgrade`.
4. At publish time, `KubernetesPublishingContext` writes placeholder entries into `values.yaml` so the chart is self-contained and `helm template` doesn't render `<no value>` for consumers who deploy outside of Aspire.

The bug was that ingress/gateway properties bypassed `AllocateParameter` and went through `ResolveExpressionAsync`, which had a degenerate fallback that returned the raw format string (`"{0}"`). This PR makes the ingress/gateway path follow the same Helm-templating + captured-values pattern as everything else. No new hybrid behavior is being introduced — the alternative would be to make ingress/gateway the one place in the publisher that resolves parameters purely in-proc, which would diverge from the rest of the chart.

Happy to fully in-proc-ify the whole publisher if that's the direction we want, but that's a much larger redesign than the ingress bug warranted, and would lose the "chart is deployable standalone via plain helm" property.

**Comment by mitchdenny** [comment]
For comparison, this is exactly what the Docker Compose publisher already does (`src/Aspire.Hosting.Docker/DockerComposeEnvironmentResource.cs`):

| Concern | Docker Compose publisher | Kubernetes/Helm publisher |
|---|---|---|
| Placeholder emitted into artifact | `${PARAM_NAME}` (native compose interpolation) | `{{ .Values.parameters.<resource>.<param> }}` (native Helm templating) |
| In-proc capture during publish | `CapturedEnvironmentVariables` | `CapturedHelmValues` |
| Where in-proc-resolved values land at deploy time | `.env.<env>` file written next to the compose file | `values.<env>.yaml` override file passed to `helm install` |
| Self-contained chart/file usable outside Aspire | Yes — `.env` placeholder file with descriptions | Yes — `values.yaml` placeholder entries |

So both publishers do "delegate to the orchestrator's native templating, *and* resolve in-proc to populate the external values file." The two designs are parallel. The ingress/gateway path in this PR is just being aligned with the existing Helm-publisher convention (which itself mirrors the Compose-publisher convention).

If we ever want to collapse to "pure in-proc, no placeholder" semantics, we'd want to do it across both publishers together rather than as a one-off in the ingress code path.

File: src/Aspire.Hosting.Kubernetes/KubernetesEnvironmentResource.cs [hunk 1]
```diff
@@ -532,6 +533,87 @@ private static async Task<string> ResolveExpressionAsync(ReferenceExpression exp
...
+            // the deploy-time override file, and ensure values.yaml has a placeholder so
+            // `helm template` (and `aspire publish` consumers that don't deploy via Aspire)
+            // can render the chart without `<no value>` substitutions.
+            var section = parameter.Secret ? HelmExtensions.SecretsKey : HelmExtensions.ParametersKey;
+            var valueKey = parameter.Name.ToHelmValuesSectionName();
+
+            if (!CapturedHelmValues.Any(c =>
+                    c.Section == section &&
+                    c.ResourceKey == owningResourceKey &&
+                    c.ValueKey == valueKey))
```
Comment (JamesNK) [inline src/Aspire.Hosting.Kubernetes/KubernetesEnvironmentResource.cs:613]: Low: When `MissingParameterValueException` is caught on the whole expression (because at least one parameter is unresolved), this handler generates Helm template references for **every** `ParameterResource` in the expression — including those that have default values and could have been inlined.

For example, if an expression combines `paramWithDefault` (has `publishValueAsDefault: true`) and `paramWithoutDefault`, both get Helm references even though the first could be resolved to its literal value here.

**Impact**: Functionally correct — `HelmDeploymentEngine` resolves all captured values at deploy time. But the published chart contains an unnecessary placeholder for parameters with known defaults, making the chart slightly less self-contained.

The fix would be to attempt `GetValueAsync` on each individual `ParameterResource` first and only fall back to a Helm reference for those that throw. However, I suspect multi-parameter expressions for ingress/gateway APIs are extremely rare in practice, so this may not be worth the added complexity.

**Review by JamesNK — APPROVED** [review]
LGTM — clean fix with good comments and test coverage. 1 low-severity observation posted (multi-parameter expression edge case).

**Comment by mitchdenny** [comment]
Thanks @JamesNK — good catch. Addressed in edf3bbbb6: `ResolveValueProviderAsync` now attempts to resolve each `ParameterResource` individually before falling back to a Helm template reference, so parameters with `publishValueAsDefault: true` get inlined even when they appear alongside an unresolved parameter in the same expression. The outer `MissingParameterValueException` from the whole-expression `GetValueAsync` call no longer "infects" the resolvable parameters.

Also picked up the rename of `WithRoute` → `WithPath` on `KubernetesIngressResource` (from #17424) which had broken the build after rebasing.
